File: typescript/test/module2/ReminderService.test.ts
import { describe, expect, it } from 'vitest';
import { fixedClock } from '../../src/module2/Clock.js';
import { LocalDate } from '../../src/module2/LocalDate.js';
import { ReminderService, type ReminderGateway } from '../../src/module2/ReminderService.js';
import { Subscription } from '../../src/module2/Subscription.js';

class RecordingReminderGateway implements ReminderGateway {
  private readonly sent: string[] = [];

  send(email: string, renewalDate: LocalDate): void {
    this.sent.push(email + '|' + renewalDate.toString());
  }

  messages(): readonly string[] {
    return [...this.sent];
  }
}

describe('ReminderServiceTest', () => {
  it('usesInjectedClockAndGateway', () => {
    const clock = fixedClock('2026-08-30T10:00:00Z');
    const gateway = new RecordingReminderGateway();
    const service = new ReminderService(clock, gateway);
    const subscription = new Subscription('developer@example.com', LocalDate.of(2026, 9, 6));

    const sent = service.sendRenewalReminder(subscription);

    expect(sent).toBe(true);
    expect(gateway.messages()).toEqual(['developer@example.com|2026-09-06']);
  });

  it('doesNotSendReminderMoreThanSevenDaysBeforeRenewal', () => {
    const clock = fixedClock('2026-08-30T10:00:00Z');
    const gateway = new RecordingReminderGateway();
    const service = new ReminderService(clock, gateway);
    const subscription = new Subscription('developer@example.com', LocalDate.of(2026, 9, 7));

    const sent = service.sendRenewalReminder(subscription);

    expect(sent).toBe(false);
    expect(gateway.messages()).toEqual([]);
  });

  it('preservesCurrentBehaviorForPastRenewalDate', () => {
    const clock = fixedClock('2026-08-30T10:00:00Z');
    const gateway = new RecordingReminderGateway();
    const service = new ReminderService(clock, gateway);
    const subscription = new Subscription('developer@example.com', LocalDate.of(2026, 8, 29));

    const sent = service.sendRenewalReminder(subscription);

    expect(sent).toBe(true);
    expect(gateway.messages()).toEqual(['developer@example.com|2026-08-29']);
  });
});

File: csharp/tests/Training.Module2.Tests/ReminderServiceTest.cs
using System.Globalization;
using Microsoft.Extensions.Time.Testing;

namespace Training.Module2.Tests;

public sealed class ReminderServiceTest
{
    [Fact]
    public void UsesInjectedClockAndGateway()
    {
        FakeTimeProvider clock = new(
            DateTimeOffset.Parse("2026-08-30T10:00:00Z", CultureInfo.InvariantCulture));
        RecordingReminderGateway gateway = new();
        ReminderService service = new(clock, gateway);
        Subscription subscription = new(
            "developer@example.com",
            new DateOnly(2026, 9, 6));

        bool sent = service.SendRenewalReminder(subscription);

        Assert.True(sent);
        Assert.Equal(
            ["developer@example.com|2026-09-06"],
            gateway.Messages());
    }

    [Fact]
    public void DoesNotSendReminderMoreThanSevenDaysBeforeRenewal()
    {
        FakeTimeProvider clock = new(
            DateTimeOffset.Parse("2026-08-30T10:00:00Z", CultureInfo.InvariantCulture));
        RecordingReminderGateway gateway = new();
        ReminderService service = new(clock, gateway);
        Subscription subscription = new(
            "developer@example.com",
            new DateOnly(2026, 9, 7));

        bool sent = service.SendRenewalReminder(subscription);

        Assert.False(sent);
        Assert.Empty(gateway.Messages());
    }

    [Fact]
    public void PreservesCurrentBehaviorForPastRenewalDate()
    {
        FakeTimeProvider clock = new(
            DateTimeOffset.Parse("2026-08-30T10:00:00Z", CultureInfo.InvariantCulture));
        RecordingReminderGateway gateway = new();
        ReminderService service = new(clock, gateway);
        Subscription subscription = new(
            "developer@example.com",
            new DateOnly(2026, 8, 29));

        bool sent = service.SendRenewalReminder(subscription);

        Assert.True(sent);
        Assert.Equal(
            ["developer@example.com|2026-08-29"],
            gateway.Messages());
    }

    private sealed class RecordingReminderGateway : ReminderService.IReminderGateway
    {
        private readonly List<string> messages = [];

        public void Send(string email, DateOnly renewalDate) =>
            messages.Add(email + "|" + renewalDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture));

        public IReadOnlyList<string> Messages() => [.. messages];
    }
}
